openerp/addons/base/ir/ir_token.py

```diff
@@ -0,0 +1,103 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import uuid
+from datetime import timedelta
+from urlparse import urljoin
+
+from odoo import _, api, fields, models
+from odoo.osv import expression
+from odoo.exceptions import AccessError
+
+
+class IrToken(models.TransientModel):
+
+    _name = 'ir.token'
+    _rec_name = 'value'
+
+    @api.model
+    def _get_random_token(self):
+        """Generate a 32 char long pseudo-random string of digits
+
+        UUID4 makes the chance of a collision (unicity constraint) highly
+        unlikely due to the expiration period of 24 hours.
+        """
+        return uuid.uuid4().hex
+
+    value = fields.Char(required=True, default=_get_random_token, size=32)
+    res_model = fields.Char("Related Document Model", required=True,
+        help="Name of the model that will process the token")
+    res_id = fields.Integer("Related Document ID", required=True,
+        help="ID of the record that will process the token")
+    partner_id = fields.Many2one("res.partner", string="Target Contact",
```

Comment:
Didn't we speak about having an email field when trying to validate on an email not linked to a partner ?